drgn/helpers/linux/kallsyms.py

```diff
@@ -73,3 +84,155 @@ def load_vmlinux_kallsyms(prog: Program) -> SymbolIndex:
+    if code == "v":
+        binding = SymbolBinding.WEAK
+        kind = SymbolKind.OBJECT
+    elif code == "w":
+        binding = SymbolBinding.WEAK
+    elif code in "tT":
+        kind = SymbolKind.FUNC
+    elif code.lower() in "srbgncd":
+        kind = SymbolKind.OBJECT
+    if binding == SymbolBinding.UNKNOWN and code.isupper():
+        binding = SymbolBinding.GLOBAL
+    return binding, kind
+
+
+def _st_info_to_binding_kind(info: int) -> Tuple[SymbolBinding, SymbolKind]:
+    binding_int = info >> 4
+    STB_WEAK = 2
+    STB_GNU_UNIQUE = 10
+    if binding_int <= STB_WEAK or binding_int == STB_GNU_UNIQUE:
+        binding = SymbolBinding(binding_int + 1)
+    else:
+        binding = SymbolBinding.UNKNOWN
+    type_ = info & 0xF
+    STT_TLS = 6
+    STT_GNU_IFUNC = 10
+    if type_ <= STT_TLS or type_ == STT_GNU_IFUNC:
+        kind = SymbolKind(type_)
+    else:
+        kind = SymbolKind.UNKNOWN
+    return binding, kind
+
+
+def _elf_sym_to_symbol(name: str, obj: Object, has_typetab: bool) -> Symbol:
+    # Linux likes to have the nm(1) character code for its symbols, which it
+    # refers to as the symbol's "type" (this is of course distinct from the ELF
+    # notion of a symbol type, let alone what drgn considers a "type"...).
+    #
+    # Prior to 5439c985c5a8 ("module: Overwrite st_size instead of st_info"),
+    # merged in v5.0, the kernel simply overwrote the "st_info" field with a
+    # single-character code that represents the nm(1) character code for that
+    # symbol. However, starting with that commit, it was switched to overwrite
+    # the "st_size" field instead! This was thankfully fixed in v5.2 with
+    # 1c7651f43777 ("kallsyms: store type information in its own array").
+    #
+    # Unfortunately, this leaves us with three possibilities:
+    # 1. Pre-v5.0: interpret the "st_info" as a character from nm(1) and try to
+    #    infer the kind and bindings.
+    # 2. 5.0-5.2: interpret the "st_info" as normal, but ignore the "st_size"
+    #    field since it is bogus.
+    # 3. 5.2+: both fields are valid, and the nm(1) code is stored in "typetab".
+    #
+    # Case 3 can be determined easily by the presence of "typetab" in "struct
+    # mod_kallsyms". However, cases 1 & 2 are indistinguishable. For our
+    # purposes, it makes more sense to fall back to case 1. After all, neither
+    # 5.0 or 5.1 were LTS kernels, nor are they actively used by any major
+    # distro. We have no way to deal with 5.0 or 5.1, whereas we can make some
+    # informed guesses for pre-5.0 based on the nm(1) code.
+    if has_typetab:
+        binding, kind = _st_info_to_binding_kind(obj.st_info.value_())
+    else:
+        binding, kind = _nm_type_to_binding_kind(chr(obj.st_info.value_()))
+    return Symbol(  # type: ignore
+        name,
+        obj.st_value.value_(),
+        obj.st_size.value_(),
+        binding,
+        kind,
+    )
+
+
+def module_kallsyms(module: Object) -> List[Symbol]:
```

Comment:
You mentioned that you wouldn't mind removing this from the public API for now, so let's do that unless you've changed your mind.